Add Elevator and Intake Subsystems

UserC2 opened the pull request:
Fixes #30 
Fixes #44

Also adds Intake subsystem, see #47

UserC2 committed:
Move ArmFeedForward to utils

UserC2 committed:
Capitalization error

UserC2 committed:
Add simple elevator feedforward

UserC2 committed:
Add WPILib's elevator feed forward (but with Sendable support)

UserC2 committed:
Add Elevator subsystem

UserC2 committed:
Control elevator subsystem

UserC2 committed:
Add elevator and intake to visualizer

UserC2 committed:
Change back to driver controller for elevator control

UserC2 committed:
Basic elevator tuning

UserC2 committed:
Remove ElevatorFeedForward with Sendable support

UserC2 commented on src/main/java/frc/robot/subsystems/Elevator.java:
From Greg:
* Don't use position/velocity conversions (do them yourself on the RIO) due to history of them not actually working.
* ^ make sure to scale upper/lower soft stops given to Spark MAX appropriately

newjanson replied:
Make sure conversions are applied in getHeight() and setHeight()

UserC2 committed:
Make simulation more accurate

UserC2 committed:
Apply position and velocity conversions with our code

gregzetko commented:
Looks good! Ready for testing once hardware is ready

UserC2 commented:
See #44! Closed until fixed!

UserC2 committed:
Fix #Elevator moves in one stage, not two! #44

UserC2 commented:
Fixed

UserC2 committed:
Move positions out of elevator and add atTargetHeight function

UserC2 committed:
Add MoveElevatorToPosition command

UserC2 committed:
Merge branch 'main' into elevator

UserC2 committed:
Comment out unused imports missed during the merge

UserC2 committed:
Fix Implement Intake Subsystem #24

UserC2 committed:
Fix Display all mechanisms (and a note in the intake) in AdvantageScope using Visualizer class #36

UserC2 committed:
Fix Add Intake Commands #31

UserC2 committed:
Wrap intake sensor in function (because it is true when there is no note in the intake)

UserC2 committed:
Track whether we have the note or not when the sensor isn't active (for visualization purposes)

UserC2 committed:
Un-swap motor direction

UserC2 committed:
Move intake commands to the intake subsystem as factory functions

Also, return the ParallelRaceGroup that ExpelNote and FeedNote get added to by 'waitUntil' so you can add them to command compositions without crashing the code.
(Previously, ExpelNote and FeedNote added themselves to a command composition but returned the raw command, so attempting to add them to another composition threw an exception.)

UserC2 committed:
Add support for intake sensor during simulation

UserC2 committed:
Merge pull request #47 from frc2609/intake

Add intake subsystem

UserC2 committed:
Set elevator inversion

UserC2 committed:
Merge branch 'main' into elevator
